[BUG] npx/npm exec don't work in child workspace when using --no/--no-install and --package

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I am using NPM workspaces. I am running a command of the format:

```
npm exec --no --workspace=my-workspace --package @graphql-codegen/cli -- graphql-codegen --config codegen.js
```

It fails with a rather terse `npm ERR! canceled`.

This is related to a previous issue #2826. The bug was partially resolved by @ruyadorno in https://github.com/npm/libnpmexec/pull/2. However, looking at the commit, it seems that the fix was only applied inside the `if (needPackageCommandSwap)` block. This only executes if no `--package` option has been specified.

If `--package` is specified, then the code falls through to https://github.com/npm/libnpmexec/blob/a9423e441b914571ec18ad8a4496a3ecf2f500e2/lib/index.js#L153. It will then try to install the package (even though it's already available locally). However, if `--no[-install]` is specified, it will simply throw an exception instead.


### Expected Behavior

The `npm exec` and `npx` commands should work when used in a workspace with the `--package` and `--no[-install]` options.

### Steps To Reproduce

_No response_

### Environment

- OS: Windows
- Node: 16.4.0
- npm: 7.19.1

I moved my original project to pnpm in the end, so I reproduced this problem by:

- Cloning https://github.com/Quramy/npm-ts-workspaces-example
- `cd packages/x-cli`
- `npm install --save-dev @graphql-codegen/cli`
- `npm exec --no --workspace=@quramy/x-cli --package @graphql-codegen/cli -- graphql-codegen`

I still see the same output:

```
npm ERR! canceled
```

--- a/lib/commands/exec.js
+++ b/lib/commands/exec.js
@@ -48,10 +48,8 @@ class Exec extends BaseCommand {
   static ignoreImplicitWorkspace = false
   static isShellout = true
 
-  async exec (_args, { locationMsg, path, runPath } = {}) {
-    if (!path) {
-      path = this.npm.localPrefix
-    }
+  async exec (_args, { locationMsg, runPath } = {}) {
+    const path = this.npm.localPrefix
 
     if (!runPath) {
       runPath = process.cwd()
@@ -95,7 +93,7 @@ class Exec extends BaseCommand {
 
     for (const path of this.workspacePaths) {
       const locationMsg = await getLocationMsg({ color, path })
-      await this.exec(args, { locationMsg, path, runPath: path })
+      await this.exec(args, { locationMsg, runPath: path })
     }
   }
 }

--- a/test/lib/commands/exec.js
+++ b/test/lib/commands/exec.js
@@ -1128,10 +1128,11 @@ t.test('forward legacyPeerDeps opt', async t => {
   )
 })
 
-t.test('workspaces', t => {
+t.test('workspaces', async t => {
   npm.localPrefix = t.testdir({
     node_modules: {
       '.bin': {
+        a: '',
         foo: '',
       },
     },
@@ -1159,68 +1160,119 @@ t.test('workspaces', t => {
   })
 
   PROGRESS_IGNORED = true
-  npm.localBin = resolve(npm.localPrefix, 'node_modules/.bin')
+  npm.localBin = resolve(npm.localPrefix, 'node_modules', '.bin')
 
-  t.test('with args, run scripts in the context of a workspace', async t => {
-    await exec.execWorkspaces(['foo', 'one arg', 'two arg'], ['a', 'b'])
+  // with arg matching existing bin, run scripts in the context of a workspace
+  await exec.execWorkspaces(['foo', 'one arg', 'two arg'], ['a', 'b'])
 
-    t.match(RUN_SCRIPTS, [
-      {
-        pkg: { scripts: { npx: 'foo' } },
-        args: ['one arg', 'two arg'],
-        banner: false,
-        path: process.cwd(),
-        stdioString: true,
-        event: 'npx',
-        env: {
-          PATH: [npm.localBin, process.env.PATH].join(delimiter),
-        },
-        stdio: 'inherit',
+  t.match(RUN_SCRIPTS, [
+    {
+      pkg: { scripts: { npx: 'foo' } },
+      args: ['one arg', 'two arg'],
+      banner: false,
+      path: npm.localPrefix,
+      stdioString: true,
+      event: 'npx',
+      env: {
+        PATH: [npm.localBin, process.env.PATH].join(delimiter),
       },
-    ])
-  })
+      stdio: 'inherit',
+    },
+    {
+      pkg: { scripts: { npx: 'foo' } },
+      args: ['one arg', 'two arg'],
+      banner: false,
+      path: npm.localPrefix,
+      stdioString: true,
+      event: 'npx',
+      env: {
+        PATH: [npm.localBin, process.env.PATH].join(delimiter),
+      },
+      stdio: 'inherit',
+    },
+  ], 'should run with multiple args across multiple workspaces')
 
-  t.test('no args, spawn interactive shell', async t => {
-    CI_NAME = null
-    process.stdin.isTTY = true
+  // clean up
+  RUN_SCRIPTS.length = 0
 
-    await exec.execWorkspaces([], ['a'])
+  // with packages, run scripts in the context of a workspace
+  config.package = ['foo']
+  config.call = 'foo'
+  config.yes = false
 
-    t.strictSame(LOG_WARN, [])
-    t.strictSame(
-      npm._mockOutputs,
+  ARB_ACTUAL_TREE[npm.localPrefix] = {
+    children: new Map([['foo', { name: 'foo', version: '1.2.3' }]]),
+  }
+
+  await exec.execWorkspaces([], ['a', 'b'])
+
+  // path should point to the workspace folder
+  t.match(RUN_SCRIPTS, [
+    {
+      pkg: { scripts: { npx: 'foo' } },
+      args: [],
+      banner: false,
+      path: resolve(npm.localPrefix, 'packages', 'a'),
+      stdioString: true,
+      event: 'npx',
+      stdio: 'inherit',
+    },
+    {
+      pkg: { scripts: { npx: 'foo' } },
+      args: [],
+      banner: false,
+      path: resolve(npm.localPrefix, 'packages', 'b'),
+      stdioString: true,
+      event: 'npx',
+      stdio: 'inherit',
+    },
+  ], 'should run without args in multiple workspaces')
+
+  t.match(ARB_CTOR, [
+    { path: npm.localPrefix },
+    { path: npm.localPrefix },
+  ])
+
+  // no args, spawn interactive shell
+  CI_NAME = null
+  config.package = []
+  config.call = ''
+  process.stdin.isTTY = true
+
+  await exec.execWorkspaces([], ['a'])
+
+  t.strictSame(LOG_WARN, [])
+  t.strictSame(
+    npm._mockOutputs,
+    [
       [
-        [
-          `\nEntering npm script environment in workspace a@1.0.0 at location:\n${resolve(
-            npm.localPrefix,
-            'packages/a'
-          )}\nType 'exit' or ^D when finished\n`,
-        ],
+        `\nEntering npm script environment in workspace a@1.0.0 at location:\n${resolve(
+          npm.localPrefix,
+          'packages/a'
+        )}\nType 'exit' or ^D when finished\n`,
       ],
-      'printed message about interactive shell'
-    )
+    ],
+    'printed message about interactive shell'
+  )
 
-    npm.color = true
-    flatOptions.color = true
-    npm._mockOutputs.length = 0
-    await exec.execWorkspaces([], ['a'])
+  npm.color = true
+  flatOptions.color = true
+  npm._mockOutputs.length = 0
+  await exec.execWorkspaces([], ['a'])
 
-    t.strictSame(LOG_WARN, [])
-    t.strictSame(
-      npm._mockOutputs,
+  t.strictSame(LOG_WARN, [])
+  t.strictSame(
+    npm._mockOutputs,
+    [
       [
-        [
+        /* eslint-disable-next-line max-len */
+        `\u001b[0m\u001b[0m\n\u001b[0mEntering npm script environment\u001b[0m\u001b[0m in workspace \u001b[32ma@1.0.0\u001b[39m at location:\u001b[0m\n\u001b[0m\u001b[2m${resolve(
+          npm.localPrefix,
+          'packages/a'
           /* eslint-disable-next-line max-len */
-          `\u001b[0m\u001b[0m\n\u001b[0mEntering npm script environment\u001b[0m\u001b[0m in workspace \u001b[32ma@1.0.0\u001b[39m at location:\u001b[0m\n\u001b[0m\u001b[2m${resolve(
-            npm.localPrefix,
-            'packages/a'
-            /* eslint-disable-next-line max-len */
-          )}\u001b[22m\u001b[0m\u001b[1m\u001b[22m\n\u001b[1mType 'exit' or ^D when finished\u001b[22m\n\u001b[1m\u001b[22m`,
-        ],
+        )}\u001b[22m\u001b[0m\u001b[1m\u001b[22m\n\u001b[1mType 'exit' or ^D when finished\u001b[22m\n\u001b[1m\u001b[22m`,
       ],
-      'printed message about interactive shell'
-    )
-  })
-
-  t.end()
+    ],
+    'printed message about interactive shell'
+  )
 })

--- a/workspaces/libnpmexec/test/index.js
+++ b/workspaces/libnpmexec/test/index.js
@@ -121,6 +121,82 @@ t.test('local pkg, must not fetch manifest for avail pkg', async t => {
   t.equal(res, 'LOCAL PKG', 'should run local pkg bin script')
 })
 
+t.test('multiple local pkgs', async t => {
+  const foo = {
+    name: '@ruyadorno/create-foo',
+    version: '2.0.0',
+    bin: {
+      'create-foo': './index.js',
+    },
+  }
+  const bar = {
+    name: '@ruyadorno/create-bar',
+    version: '2.0.0',
+    bin: {
+      'create-bar': './index.js',
+    },
+  }
+  const path = t.testdir({
+    cache: {},
+    npxCache: {},
+    node_modules: {
+      '.bin': {},
+      '@ruyadorno': {
+        'create-foo': {
+          'package.json': JSON.stringify(foo),
+          'index.js': `#!/usr/bin/env node
+  require('fs').writeFileSync(process.argv.slice(2)[0], 'foo')`,
+        },
+        'create-bar': {
+          'package.json': JSON.stringify(bar),
+          'index.js': `#!/usr/bin/env node
+  require('fs').writeFileSync(process.argv.slice(2)[0], 'bar')`,
+        },
+      },
+    },
+    'package.json': JSON.stringify({
+      name: 'pkg',
+      dependencies: {
+        '@ruyadorno/create-foo': '^2.0.0',
+        '@ruyadorno/create-bar': '^2.0.0',
+      },
+    }),
+  })
+  const runPath = path
+  const cache = resolve(path, 'cache')
+  const npxCache = resolve(path, 'npxCache')
+
+  const setupBins = async (pkg) => {
+    const executable =
+      resolve(path, `node_modules/${pkg.name}/index.js`)
+    fs.chmodSync(executable, 0o775)
+
+    await binLinks({
+      path: resolve(path, `node_modules/${pkg.name}`),
+      pkg,
+    })
+  }
+
+  await Promise.all([foo, bar]
+    .map(setupBins))
+
+  await libexec({
+    ...baseOpts,
+    localBin: resolve(path, 'node_modules/.bin'),
+    cache,
+    npxCache,
+    packages: ['@ruyadorno/create-foo', '@ruyadorno/create-bar'],
+    call: 'create-foo resfile && create-bar bar',
+    path,
+    runPath,
+  })
+
+  const resFoo = fs.readFileSync(resolve(path, 'resfile')).toString()
+  t.equal(resFoo, 'foo', 'should run local pkg bin script')
+  const resBar = fs.readFileSync(resolve(path, 'bar')).toString()
+  t.equal(resBar, 'bar', 'should run local pkg bin script')
+})
+
 t.test('local file system path', async t => {
   const path = t.testdir({
     cache: {},

